lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,72 @@
+module ActiveModel
+  class Serializer
+    module Adapter
+      class JsonApi
+        module Deserialization
+          module_function
+
+          # Parse a Hash representing a JSON API document into an ActiveRecord-ready hash.
+          #
+          # @param [Hash] document
+          # @param [Hash] options
+          #   fields: Array of symbols and a Hash. Specify whitelisted fields, optionally
+          #     specifying the attribute name on the model.
+          # @return [Hash] ActiveRecord-ready hash
+          #
+          def parse(document, options = {})
+            fields = parse_fields(options[:fields])
+
+            hash = {}
+
+            primary_data = document.fetch('data', {})
+            hash[:id] = primary_data['id'] if primary_data['id'] && (fields.nil? || fields[:id])
+
+            hash.merge!(parse_attributes(primary_data['attributes'], fields))
+            hash.merge!(parse_relationships(primary_data['relationships'], fields))
+
+            hash
+          end
+
+          # @api private
```

Comment:
this converts options[:fields] into nil or a Hash?  Would `extract_options` work?